pkg/auth/providers/azure/clients/ms_graph_client.go

```diff
@@ -0,0 +1,303 @@
+}
+
+type authorizer struct {
+	authResult *customAuthResult
+}
+
+func (a *authorizer) AuxiliaryTokens() ([]*oauth2.Token, error) {
+	return nil, nil
+}
+
+func (a *authorizer) Token() (*oauth2.Token, error) {
+	return &oauth2.Token{
+		AccessToken: a.authResult.AccessToken,
+		TokenType:   "Bearer",
+		Expiry:      a.authResult.ExpiresOn,
+	}, nil
+}
+
+// oidFromAuthCode verifies a user.
+func oidFromAuthCode(token string, config *v32.AzureADConfig) (string, error) {
+	cred, err := confidential.NewCredFromSecret(config.ApplicationSecret)
+	if err != nil {
+		return "", fmt.Errorf("could not create a cred from a secret: %w", err)
+	}
+	confidentialClientApp, err := confidential.New(config.ApplicationID, cred, confidential.WithAuthority(fmt.Sprintf("%s%s", config.Endpoint, config.TenantID)))
+	if err != nil {
+		return "", err
+	}
+	scope := fmt.Sprintf("%s/%s", config.GraphEndpoint, ".default")
+	authResult, err := confidentialClientApp.AcquireTokenByAuthCode(context.Background(), token, config.RancherURL, []string{scope})
+	if err != nil {
+		return "", err
+	}
+
+	return authResult.IDToken.Oid, nil
+}
+
+// ClientFromAuthConfig attempts to get an access token to the Microsoft Graph API.
+// It first tries to fetch the token from the cache. If that fails, it tries to acquire it directly from the auth provider.
+func ClientFromAuthConfig(config *v32.AzureADConfig, replace cache.ExportReplace) (AzureClient, error) {
+	cred, err := confidential.NewCredFromSecret(config.ApplicationSecret)
+	if err != nil {
+		return nil, fmt.Errorf("could not create a cred from a secret: %w", err)
+	}
+	confidentialClientApp, err := confidential.New(config.ApplicationID, cred,
+		confidential.WithAccessor(replace),
+		confidential.WithAuthority(fmt.Sprintf("%s%s", config.Endpoint, config.TenantID)))
+	if err != nil {
+		return nil, err
+	}
+	scope := fmt.Sprintf("%s/%s", config.GraphEndpoint, ".default")
+
+	var ar confidential.AuthResult
+	ar, err = confidentialClientApp.AcquireTokenSilent(context.Background(), []string{scope})
+	if err != nil {
+		logrus.Infof("failed to get the access token from cache: %v", err)
+		ar, err = confidentialClientApp.AcquireTokenByCredential(context.Background(), []string{scope})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return newClient(config.TenantID, getCustomAuthResult(&ar)), nil
+}
+
+// AccessTokenCache is responsible for reading (replacing) the access token from some storage (a secret in the database,
+// in this case) and writing it (exporting) to storage. The Microsoft Graph SDK is responsible for verifying and calling
+// the cache methods when needed, so this type simply implements the interface for the cache that the SDK requires.
+type AccessTokenCache struct {
+	Secrets corev1.SecretInterface
+}
+
+func (c AccessTokenCache) Replace(cache cache.Unmarshaler, key string) {
+	secret, err := common.ReadFromSecret(c.Secrets, "cattle-global-data:azuread-access-token", "access-token")
```

Comment:
You may want to consider using `common.SecretsNamespace + ":azuread-access-token"` to stay in pairity with the way that this functions.